### ISSUE TITLE
Make WebJars support OSGi

### ISSUE THREAD

**Issue by davidvanlaatum** [webjars/webjars#1307]
Can support be added for OSGi please?

**Comment by jamesward** [comment]
Associated PRs:
https://github.com/webjars/webjars-locator/pull/94
https://github.com/webjars/webjars-locator-core/pull/9

**Comment by pleku** [comment]
It looks like OSGi support almost happened but not quite and now it is not looking likely, at least in the near future. Do you have any comments on this @jamesward ?

**Comment by jamesward** [comment]
Really I just need help since this is not something I understand.

**Comment by Sandared** [comment]
@jamesward  I think there are two different ways how webjars could support OSGi:

1) Being a plain OSGi bundle: This would merely require a MANIFEST.MF file in the META-INF folder and maybe an additional header (e.g. WebJar-Resource) so that a BundleTracker in OSGi can be notified whenever a webjar is available at runtime. This BundleTracker then can do whatever he likes to do with the content of the WebJar (e.g. registering its contents as a resource at the HttpService). This approach would be the easiest way, as all you need to do is to add the MANIFEST.MF file.

2) Being an OSGi bundle with Require-/Provide-Capability: This would be the more sophisticated version, as it would support automatic resolution in an OSGi environment. This would require all Maven dependencies to be also declared as Require-/Provide-Capabilites in the MANIFEST.MF

Both solutions are simple copy&paste actions I think. The POM would need an additional plugin declaration for the [maven-bnd-plugin](https://github.com/bndtools/bnd/blob/master/maven/bnd-maven-plugin/README.md) and an additional bnd.bnd file that contains some instructions for this plugin.

**Comment by jamesward** [comment]
Cool.  Want to try it on a WebJar and see which approach feels better?

**Comment by Sandared** [comment]
I will have a look at it at the weekend

**Comment by Sandared** [comment]
@jamesward  would this be a feasible solution? I think this needs to be done for all webjars, so this seems to be quite a lot of work to do. Maybe it could be automated partially? If not I'm willing to help with the copy&Paste ;)

**Comment by jamesward** [comment]
This seems fine.  Can you send a pull request with the changes?  Before we try to automate updating all the WebJars, it'd be great to test the solution for the Polymer one.

**Comment by Sandared** [comment]
I can do this, but I'm not sure for which repository to create a pull request. [This polymer repo](https://github.com/webjars/polymer) seems outdated. Can you please point me to the right repository?

**Comment by jamesward** [comment]
Yeah, I think most folks using Polymer via WebJars are using the Bower GitHub style WebJars.  But we should try the OSGI thing on a Classic WebJar (those manually deployed with builds on https://github.com/webjars).  Is there are Classic WebJar you use that we could try this on?

**Comment by Sandared** [comment]
I'm not using any webjar yet. I'm only following the vaadin OSGi support  development which currently is blocked by webjars not supporting OSGi ;) But I can do it with the above mentioned polymer repo if that's ok?

**Comment by jamesward** [comment]
You can add it to the Classic Polymer WebJar but it is terribly out-of-date because of the way Polymer now does distribution.  If that is ok for your testing purposes, go for it.

**Comment by Sandared** [comment]
I've created a pull request for the jquery webjar. This one's resolving in my OSGi environment now if I declare a dependency to it and has all the other necessary OSGi header in its MANIFEST.MF file. Additionally I've created a WebJars-Resource Header that is pointing to the files/resources that should be  registered at a `HttpService` in OSGi. Provide-/Require-Capabilities are introduced to enable the OSGi resolver to automatically resolve dependencies between WebJars.

**Comment by Sandared** [comment]
@jamesward is the proposed solution sufficient? Or is there anything else I can do to support you?

**Comment by jamesward** [comment]
Sorry for the delay and thanks for the PR!  I've provided feedback there.

**Comment by zhsc** [comment]
Just for clarity, are we referring to the two pull requests previously mentioned?  Those seem to be from two years ago with no updates since then. 

Thanks for the support helping the osgi community, especially coming from vaadin 7/8 users who depend on this to upgrade. 

Where's the best place to track and help resolve this seemingly last remaining blocker? 

Thanks again, sincerely.

**Comment by Sandared** [comment]
The PR we talked about ist this one https://github.com/webjars/jquery/pull/58

This led to the creation of a plugin to automatically generate the osgi version of a bundle.
https://github.com/jamesward/osgi-version-maven-plugin
This might also be the right place for further contributions :)

**Comment by zhsc** [comment]
Great thanks.  I did leave a note on this other issue too with a different approach if that helps. 

https://github.com/webjars/webjars-locator-core/pull/9#issuecomment-415921413

**Comment by jamesward** [comment]
@Sandared Can we update https://github.com/webjars/jquery/pull/58 to use the osgi-version-maven-plugin?
